File: forms/outpatient-outcome-report/front-end-form-with-svelte/src/lib/engine/oocg-grader.ts
import type { AssessmentData, GradingResult } from './types';
import { gradeClinical } from './clinical-domain';
import { gradePROM } from './prom-domain';
import { gradePREM } from './prem-domain';
import { gradeOperational } from './operational-domain';
import { detectFlaggedIssues } from './flagged-issues';
import { gradeMax, promisGphTScore, promisMhTScore } from './utils';

/**
 * Orchestrator: compute the full OOCG grading result from assessment data.
 *
 * The overall grade is the worst of the four domain grades ("highest severity wins").
 */
export function gradeOOCG(data: AssessmentData): GradingResult {
	// Update derived PROMIS T-scores before grading
	const enrichedData: AssessmentData = {
		...data,
		promPromis: {
			...data.promPromis,
			globalPhysicalHealthTScore: promisGphTScore(data.promPromis),
			globalMentalHealthTScore: promisMhTScore(data.promPromis)
		}
	};

	const clinical = gradeClinical(enrichedData);
	const prom = gradePROM(enrichedData);
	const prem = gradePREM(enrichedData);
	const operational = gradeOperational(enrichedData);

	const overallGrade = gradeMax([
		clinical.grade,
		prom.grade,
		prem.grade,
		operational.grade
	]);

	const firedRules = [
		...clinical.rules,
		...prom.rules,
		...prem.rules,
		...operational.rules
	];

	const flaggedIssues = detectFlaggedIssues(enrichedData, clinical.grade, prem.grade);

	return {
		overallGrade,
		clinicalGrade: clinical.grade,
		promGrade: prom.grade,
		premGrade: prem.grade,
		operationalGrade: operational.grade,
		firedRules,
		flaggedIssues,
		timestamp: new Date().toISOString()
	};
}

File: forms/outpatient-outcome/back-end-with-loco/src/engine/oocg_grader.rs
//! Oocg grader module.

use super::clinical_domain::grade_clinical;
use super::flagged_issues::detect_flagged_issues;
use super::operational_domain::grade_operational;
use super::prem_domain::grade_prem;
use super::prom_domain::grade_prom;
use super::types::{AssessmentData, GradingResult};
use super::utils::{grade_max, promis_gph_t_score, promis_mh_t_score};

/// Compute the full OOCG grading result from assessment data.
///
/// Overall grade = worst of the four domain grades ("highest severity wins").
pub fn grade_oocg(data: &AssessmentData) -> GradingResult {
    // Enrich PROMIS T-scores before grading (matches the front-end orchestrator).
    let mut enriched = data.clone();
    enriched.prom_promis.global_physical_health_t_score = promis_gph_t_score(&enriched.prom_promis);
    enriched.prom_promis.global_mental_health_t_score = promis_mh_t_score(&enriched.prom_promis);

    let clinical = grade_clinical(&enriched);
    let prom = grade_prom(&enriched);
    let prem = grade_prem(&enriched);
    let operational = grade_operational(&enriched);

    let overall_grade = grade_max(&[
        clinical.grade.clone(),
        prom.grade.clone(),
        prem.grade.clone(),
        operational.grade.clone(),
    ]);

    let mut fired_rules = Vec::new();
    fired_rules.extend(clinical.rules.iter().cloned());
    fired_rules.extend(prom.rules.iter().cloned());
    fired_rules.extend(prem.rules.iter().cloned());
    fired_rules.extend(operational.rules.iter().cloned());

    let flagged_issues = detect_flagged_issues(&enriched, &clinical.grade, &prem.grade);

    GradingResult {
        overall_grade,
        clinical_grade: clinical.grade,
        prom_grade: prom.grade,
        prem_grade: prem.grade,
        operational_grade: operational.grade,
        fired_rules,
        flagged_issues,
        timestamp: chrono::Utc::now().to_rfc3339(),
    }
}
